Refactoring #1

michel2323 opened the pull request:
I did a little refactoring inspired by Francois' work in ExaPF. Not yet all done I'd say, but feel free to leave comments.

frapac commented on src/Evaluators/FullModelEvaluator.jl:
```diff
@@ -1,16 +1,46 @@
+struct FullModelEvaluator <: AbstractNLPEvaluator
```
I would maybe rename as `FullSpaceModelEvaluator` ? Or just `FullSpaceModel`?

michel2323 replied:
There is also the distinction between fullspace and reduced space for the block models. But I guess that doesn't matter right?

AnirudhSubramanyam replied:
@michel2323 : yes those have different meanings in ExaPF and ProxAL. I would suggest `FullSpaceModelEvaluator`

michel2323 replied:
Fixed.

frapac commented on src/Evaluators/FullModelEvaluator.jl:
```diff
@@ -1,16 +1,46 @@
+    comm::MPI.Comm
+end
+
+function FullModelEvaluator(case_file::String, load_file::String,
```
I would maybe try to get a more consistent indentation. Like 
```julia
function FullModelEvaluator(
    case_file::String, load_file::String,
    modelinfo::ModelParams,
    algparams::AlgParams,
    space::AbstractSpace=FullSpace(),
    comm::MPI.Comm = MPI.COMM_WORLD
)

michel2323 replied:
I am not sure I know what the consistent way is :-) . Do you have a rule of thumb?

frapac replied:
I am trying to follow that guideline: https://jump.dev/JuMP.jl/v0.21.1/style/

michel2323 replied:
Fixed here and in other places.

frapac commented on src/Evaluators/FullModelEvaluator.jl:
```diff
@@ -1,16 +1,46 @@
+                     comm::MPI.Comm = MPI.COMM_WORLD)
+    rawdata = RawData(case_file, load_file)
+    opfdata = opf_loaddata(rawdata;
+                        time_horizon_start = 1,
```
Why the start of the time horizon is not an argument of the function?

michel2323 replied:
I am not sure it should be an argument. @AnirudhSubramanyam What is your take? I did not see a case where this was different from 1.

AnirudhSubramanyam replied:
@frapac : good catch! @michel2323 : we don't have such a case now, but I can envision its use when we try to run this within a model predictive control application

michel2323 replied:
Fixed

frapac commented on src/Evaluators/FullModelEvaluator.jl:
```diff
@@ -1,16 +1,46 @@
+            modelinfo)
+
+    # ctgs_arr = deepcopy(rawdata.ctgs_arr)
+    alminfo = ProxALMData(opfdata, rawdata, modelinfo, algparams, space, false)
```
Why do we set `false` here?

michel2323 replied:
I am not sure I understand what this does. If this is set to `true` it will execute this in the constructor:

frapac replied:
I confess I don't know either...

AnirudhSubramanyam replied:
@frapac @michel2323 that exists just so we can solve the full space model without decomposition, and use that as a reference solution to track convergence of the algorithm.

michel2323 replied:
Ok, I fixed this now. I think this was dead code as the constructor also solved the full model before. Now this is called explicitly with `optimize!`. I hope I didn't break anything. @frapac @AnirudhSubramanyam

michel2323 replied:
@AnirudhSubramanyam This code was never called in the tests. The `FullSpaceEvaluator` now runs with whatever is set by the user. Should I override the algorithm parameters with `algparams.mode = :nondecomposed` and `algparams.mode = :lyapunov_bound`. Because that is now removed.

AnirudhSubramanyam replied:
@michel2323 I'm not sure if we should override it. Or perhaps we give a warning before overriding it..? Even if we override, its not clear if we should go with `algparams.mode = :nondecomposed` or `algparams.mode = :lyapunov_bound`. So maybe let's just leave it as it is.

michel2323 replied:
That makes sense.

frapac commented on src/Evaluators/ProxALEvalutor.jl:
```diff
@@ -0,0 +1,256 @@
+The argument `space` specifies whether we are working
+in the `FullSpace()` or in the `ReducedSpace()`.
+"""
+function ProxALEvaluator(case_file::String, load_file::String,
```
Indentation?

michel2323 replied:
Fixed

frapac commented on src/Evaluators/ProxALEvalutor.jl:
```diff
@@ -0,0 +1,256 @@
+struct ProxALEvaluator <: AbstractNLPEvaluator
```
I think the meaning behind the name `ProxALEvaluator` is not straightforward to grasp

michel2323 replied:
I don't know the actual name of the method either.

michel2323 replied:
Maybe we can solve that a later stage.

frapac replied:
sounds good to me!

frapac commented on src/Evaluators/ProxALEvalutor.jl:
```diff
@@ -0,0 +1,256 @@
+                        time_horizon_end = modelinfo.num_time_periods,
+                        load_scale = modelinfo.load_scale,
+                        ramp_scale = modelinfo.ramp_scale)
+    set_rho!(algparams,
```
Maybe rename `set_rho!` as `set_penalty!` ?

michel2323 replied:
@AnirudhSubramanyam What do you think?

AnirudhSubramanyam replied:
@michel2323 I agree with @frapac `set_penalty!` is the better name.

michel2323 replied:
Fixed

frapac commented on src/Evaluators/ProxALEvalutor.jl:
```diff
@@ -0,0 +1,256 @@
+    function transfer!(blk, opt_sol, solution)
+        # Pg
+        fr = 1 ; to = ngen * k_per_block
+        opt_sol[fr:to, blk] .= solution.pg[:]
```
I am all in to find a solution that does not depend on this explicit indexing of the vector `opt_sol`

frapac replied:
(not related to this PR though)

michel2323 replied:
Yes. It's due to the ad-hoc MPI implementation.

frapac reviewed (changes requested):
Looks good to me! 
The only thing I am not sure with is the naming: I would maybe replace `Evaluator` by `Model`, but I am open to the discussion :)

michel2323 commented:
I addressed all your comments. Thank you! @frapac @AnirudhSubramanyam . The open comments I am either a bit unsure or they can be addressed at a later stage. Let me know if it's ready to rebase and merge.

frapac reviewed (approved):
I think we are ready to merge this PR!
